pageserver/src/relish_storage/synced_storage.rs

```diff
@@ -0,0 +1,70 @@
+use std::sync::mpsc::Receiver;
+use std::time::Duration;
+use std::{collections::BinaryHeap, sync::Mutex, thread};
+
+use zenith_utils::zid::{ZTenantId, ZTimelineId};
+
+use crate::{relish_storage::RelishStorage, PageServerConf};
+
+lazy_static::lazy_static! {
+    static ref UPLOAD_QUEUE: Mutex<BinaryHeap<SyncTask>> = Mutex::new(BinaryHeap::new());
+}
+
+pub struct StorageAccessor {
+    pub storage_uploader: StorageUploader,
+    pub download_signal: Receiver<(ZTenantId, ZTimelineId)>,
+}
+
+pub struct StorageUploader;
+
+impl StorageUploader {
+    // pub fn upload_relish(&self, local_timeline: LocalTimeline) {
+    //     UPLOAD_QUEUE
+    //         .lock()
+    //         .unwrap()
+    //         .push(SyncTask::Upload(local_timeline))
+    // }
+}
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+enum SyncTask {}
+
+pub fn run_storage_sync_thread<
+    P: std::fmt::Debug,
+    S: 'static + RelishStorage<RelishStoragePath = P>,
+>(
+    config: &'static PageServerConf,
+    relish_storage: S,
+    max_concurrent_sync: usize,
+) -> anyhow::Result<Option<(StorageAccessor, thread::JoinHandle<anyhow::Result<()>>)>> {
+    let runtime = tokio::runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()?;
+    let (sender, receiver) = std::sync::mpsc::channel();
+    let handle = thread::Builder::new()
+        .name("Queue based relish storage sync".to_string())
+        .spawn(move || loop {
+            let mut queue_accessor = UPLOAD_QUEUE.lock().unwrap();
+            log::debug!("Upload queue length: {}", queue_accessor.len());
+            let next_task = queue_accessor.pop();
+            drop(queue_accessor);
+            match next_task {
+                Some(task) => runtime.block_on(async {
+                    // suppress warnings
+                    let _ = (config, task, &sender, &relish_storage, max_concurrent_sync);
+                    todo!("omitted for brevity")
```

Comment:
At some point here, the `sender` gets its message on upload: https://github.com/zenithdb/zenith/pull/686/files#diff-8ef63341d08bdfccc129283af160e1b6032defcd0bc929974510710cdf3d6478R314
(you might need to load diff for `synced_storage.rs` first to make the link work)

`sender` is moved into the thread loop and gets dropped with the thread stop.